### PR TITLE
feat(hermes): add RCA synthetic/e2e suite + session evidence tooling

### LINKED ISSUE #1510: Hermes incident-identification scenarios 2/5 — Long-running/agentic reliability (compression, KV cache, CLI hangs, cron, ghost sessions)
## Problem statement

Once a Hermes session runs for more than a few minutes — context compression, KV cache refresh, cron triggers, TUI continuations — the failure modes shift from "the provider returned 4xx" to "the agent is hung, ghosted, or silently corrupted". These are the failures users notice last and trust least, because the agent *appears* to be working.

OpenSRE today has no scenarios that exercise "the agent has been stuck for 20 minutes" or "the cron job produced output but never delivered it". This issue adds 5 long-running / agentic-reliability scenarios as both synthetic (fixture-backed, no LLM key) and e2e (captured Hermes session) tests.

## What Hermes is, and why openSRE needs this coverage

Hermes is a long-running coding/agent CLI with a sprawling adapter layer: ~19 LLM provider labels, ~13 messaging adapters, ~7 execution backends (Docker, SSH, Modal, Daytona, Vercel, Singularity, local). Its upstream tracker carries 1000+ open issues. Production failures cluster into three families: *adapter-level* (HTTP 4xx/5xx, malformed SSE, header drops, IMDS auto-detect overrides), *agent-level* (compression corruption, KV-cache thrash, hung interrupt queue, ghost sessions), and *capability-gap-level* (no real multi-agent orchestration, no determinism engine, no audit trail, no RBAC).

When one of these failures fires, the user sees an alert like "Hermes session crashed" or "Hermes cron run produced no delivery" and asks openSRE to identify the root cause. openSRE has rich e2e + synthetic coverage for K8s, RDS, Lambda, Prefect, and Apache Flink — but **nothing for "the AI agent itself failed"**. This 5-part series fills the gap.

## Harness this issue requires

Scenarios live under two trees that mirror existing openSRE conventions:

- `tests/synthetic/hermes/` — fixture-backed, no-network, no-LLM-key tests. Mirrors `tests/synthetic/eks/`.
- `tests/e2e/hermes/` — captured-real-session tests. Mirrors `tests/e2e/openclaw/` (issue #1484).

Both rely on a `FixtureHermesBackend` (under `tests/synthetic/mock_hermes_backend/`) that wraps captured Hermes session data into the same envelope a future production `HermesBackend` would return, so investigation tools can read the captured evidence without touching the real Hermes process. Mirrors `FixtureEKSBackend` in `tests/synthetic/mock_eks_backend/backend.py`.

Each synthetic scenario is a directory `tests/synthetic/hermes/NNN-<failure-mode>/` containing:

- `scenario.yml` — metadata (`scenario_id`, `failure_mode`, `severity`, `available_evidence`, `scenario_difficulty`). Same shape as `tests/synthetic/eks/000-healthy/scenario.yml`.
- `alert.json` — the inbound alert openSRE receives, including `commonAnnotations.context_sources: "hermes"`.
- `answer.yml` — expected `root_cause_category`, `required_keywords`, `forbidden_categories`, `optimal_trajectory`, `max_investigation_loops`. Same shape as `tests/synthetic/eks/001-oomkilled-crashloop/answer.yml`.
- One JSON file per evidence source declared in `scenario.yml:available_evidence` (schemas defined below).

The synthetic scenario loader, run-suite, and scorer mirror `tests/synthetic/eks/scenario_loader.py`, `run_suite.py`, and `test_suite.py` — same shape, Hermes-specific vocabulary defined in a new `tests/synthetic/hermes/hermes_schemas.py`.

E2E scenarios follow `tests/AGENTS.md` (failure-first design): boot or load a real Hermes session that exhibits the failure, capture full failure context, build an alert via the `create_alert` factory, invoke the openSRE pipeline inside a `@traceable` block, assert on `root_cause_category` + `validity_score > 0.7`.

**If the harness files don't yet exist when you start this issue** (`hermes_schemas.py`, `scenario_loader.py`, `run_suite.py`, `test_suite.py`, `tests/synthetic/mock_hermes_backend/backend.py`, the baseline `000-healthy` scenario, the `Makefile` `test-hermes` target, the `pyproject.toml` `hermes` marker), **add them as part of this PR** — they are shared across the 5-part series and the first PR to merge owns the harness. Coordinate via the cross-linked issues at the bottom.

## Evidence sources used by this issue

Define in `tests/synthetic/hermes/hermes_schemas.py` (extending whatever the harness already declares):

```python
VALID_HERMES_EVIDENCE_SOURCES |= frozenset({
    "hermes_session_log",        # JSONL of agent events (also used by part 1/5)
    "hermes_message_history",    # full ordered message list (system/user/assistant/tool_call/tool)
    "hermes_kv_cache_state",     # cache hit/miss counters, last-cached prefix, last-invalidated reason
    "hermes_runtime_state",      # extended with interrupt_queue_depth, last_progress_ts, is_blocked
    "hermes_cron_state",         # cron schedule + per-run timing + last_delivery_status
    "hermes_session_topology",   # session_id, parent_session_id, continuation_session_id (ghost detection)
})

VALID_HERMES_FAILURE_MODES |= frozenset({
    "agent_state_corruption",    # compression broke message ordering
    "agent_hang",                # interrupt queue drain removed; tool call blocks forever
    "delivery_hang",             # cron produced output, never delivered
    "performance_degradation",   # KV cache thrash from message-format drift
    "ghost_session",             # TUI compression continuation forked an invisible session
})
```

Per-source JSON shape (define `TypedDict` + `validate_*` for each, matching `k8s_schemas.py`):

- **`hermes_message_history.json`** — `{ "session_id": str, "messages": [ { "role": "system|user|assistant|tool_call|tool", "content": str | object, "tool_call_id": str | null, "ts": str } ] }`. The agent must be able to detect invalid `tool_call` → `tool` ordering (every `tool` message must reference a prior `tool_call_id`).
- **`hermes_kv_cache_state.json`** — `{ "session_id": str, "cache_hits": int, "cache_misses": int, "last_cached_prefix_bytes": int, "last_invalidated_reason": str, "messages_with_cache_miss": [ { "message_index": int, "expected_prefix_bytes": int, "actual_prefix_bytes": int, "diff_kind": "whitespace|role_format|content" } ] }`. The `diff_kind` field is what makes message-format-drift diagnosable.
- **`hermes_runtime_state.json`** — `{ "pid": int, "started_at": str, "frozen_now_ts": str, "interrupt_queue_depth": int, "last_progress_ts": str, "is_blocked": bool, "blocking_call": { "tool_name": str, "started_at": str, "duration_s": int } | null }`. Every scenario sets `frozen_now_ts` so the `is_blocked = (frozen_now_ts - last_progress_ts) > hang_threshold_s` calculation is deterministic — no wall-clock time in tests.
- **`hermes_cron_state.json`** — `{ "schedule_cron": str, "last_run": { "started_at": str, "agent_completed_at": str | null, "delivery_started_at": str | null, "delivery_completed_at": str | null, "delivery_status": "ok|hung|failed|never_started" } }`. The "agent completed but delivery never started" pattern is the diagnostic anchor for #5209.
- **`hermes_session_topology.json`** — `{ "visible_session_id": str, "all_sessions": [ { "session_id": str, "parent_session_id": str | null, "continuation_of": str | null, "last_message_ts": str, "message_count": int } ] }`. The agent must follow `continuation_of` from the visible session to discover the ghost.

## Common conventions

- Pytest markers: `@pytest.mark.synthetic` for synthetic scenarios, `@pytest.mark.e2e` for e2e.
- All fixtures must be free of real provider tokens or user PII — scrub via `app/masking/` before committing.
- Score model: `tests/synthetic/eks/run_suite.py:score_result` is the reference. Failure-first design from `tests/AGENTS.md`.
- Alert factory: use `create_alert` with `context_sources="hermes"` plus issue-specific annotations.
- Determinism: all hang/timing scenarios MUST use `frozen_now_ts` from the runtime state fixture, never wall-clock time.

## Scenarios to add

- [ ] **`010-compression-invalid-tool-ordering`** — Context compression emits an invalid `tool_call` / `tool` message ordering, the next provider call returns HTTP 400, the session crashes mid-conversation. Upstream Hermes #20883. Expected RCA: `root_cause_category: agent_state_corruption`, names "context compression", identifies the malformed message sequence, points at the compressor's invariant violation. Fixture must include both pre- and post-compression `hermes_message_history` so the agent can compare.
- [ ] **`011-cli-hang-no-interrupt-drain`** — CLI hangs indefinitely when the agent thread blocks because the `interrupt_queue` drain was removed in v0.12.0. Upstream Hermes #20271. Expected RCA: `root_cause_category: agent_hang`, identifies the missing interrupt path, names the version regression, recommends the v0.12.0 → v0.12.1 workaround. `hermes_runtime_state.is_blocked: true` is the anchor.
- [ ] **`012-cron-hang-post-output`** — Cron-triggered Hermes job produces full agent output but never reaches the delivery step (Slack/Telegram post). Upstream Hermes #5209. Expected RCA: `root_cause_category: delivery_hang`, identifies the post-output delivery hang, distinguishes from "agent failed", names which delivery adapter never ran. `hermes_cron_state.last_run.delivery_status: never_started` is the anchor.
- [ ] **`013-kv-cache-invalidation-format-drift`** — KV cache is invalidated on every new user message because the formatted message bytes differ even when the semantic content does not (whitespace / role-format drift). Upstream Hermes #4555. Expected RCA: `root_cause_category: performance_degradation`, identifies the cache-miss pattern, names the formatting-drift root cause, suggests the canonicalization fix. `hermes_kv_cache_state.messages_with_cache_miss[*].diff_kind` is the anchor.
- [ ] **`014-tui-compression-ghost-session`** — TUI compression continuation forks a "ghost" session that diverges from the visible session; later messages land in the ghost, user sees no progress. Upstream Hermes #20001. Expected RCA: `root_cause_category: ghost_session`, identifies the dual-session split, names the compression-continuation handler, points at the session-id reconciliation gap. The alert annotations include only the visible `session_id`; the agent must follow `hermes_session_topology.continuation_of` to discover the ghost.

## Test layout

```
tests/e2e/hermes/long_running/
├── __init__.py
├── test_compression_invalid_ordering.py
├── test_cli_hang_no_interrupt.py
├── test_cron_hang_post_output.py
├── test_kv_cache_invalidation.py
├── test_tui_compression_ghost_session.py
└── fixtures/                                # captured Hermes sessions + frozen runtime snapshots

tests/synthetic/hermes/
├── 010-compression-invalid-tool-ordering/
├── 011-cli-hang-no-interrupt-drain/
├── 012-cron-hang-post-output/
├── 013-kv-cache-invalidation-format-drift/
└── 014-tui-compression-ghost-session/
```

If the harness directories `tests/synthetic/hermes/`, `tests/e2e/hermes/`, and `tests/synthetic/mock_hermes_backend/` don't yet exist, create them per the **Harness this issue requires** section above.

## Test flow specifics

- **Hang detection**: each hang scenario captures a snapshot taken N seconds after `last_progress_ts`. The synthetic backend exposes `is_blocked = (frozen_now_ts - last_progress_ts) > hang_threshold_s`. The investigation must correctly distinguish a hang from a long-running-but-progressing tool call.
- **Compression bugs**: the synthetic fixture must include both the *pre-compression* and *post-compression* `hermes_message_history` so the agent can compare and identify the corruption, not just observe the eventual 400.
- **Ghost sessions**: the alert annotations include only the visible `session_id`; the agent must follow `continuation_of` to discover the ghost.
- **Delivery hang**: assert the report names *both* the agent-side success (output produced) and the delivery-side failure (delivery never started) — two findings, not one.

## Acceptance criteria

- [ ] All 5 scenarios added with both an e2e test and a synthetic scenario fixture.
- [ ] `hermes_message_history`, `hermes_kv_cache_state`, `hermes_cron_state`, `hermes_session_topology` evidence sources defined in `tests/synthetic/hermes/hermes_schemas.py` (with `TypedDict` + `validate_*`).
- [ ] Each hang scenario asserts the agent reports `root_cause_category: agent_hang` (not `unknown`, not `provider_transport_error`).
- [ ] The compression scenario asserts the report distinguishes "the provider rejected the request" from "we sent a malformed sequence to the provider".
- [ ] Every timing-sensitive scenario uses `frozen_now_ts` from the fixture; no wall-clock time in tests.
- [ ] Captured fixtures pass a `detect-secrets` scan in CI.
- [ ] `make lint`, `make typecheck`, `make test-cov`, and `make test-hermes` all pass.

## Risks

- Hang detection is timing-sensitive; the synthetic backend must use a fixed `frozen_now_ts` per scenario rather than wall-clock time so tests are deterministic.
- Compression-bug fixtures are large (full message histories); cap each fixture at < 50KB and link to a longer external sample if needed.

## References

- Upstream Hermes issues: #20883, #20271, #5209, #4555, #20001.
- Existing synthetic harness pattern: `tests/synthetic/eks/` (scenario loader, run suite, mock backend).
- Existing e2e pattern: `tests/e2e/openclaw/` (target layout from #1484).
- `tests/AGENTS.md` — `@traceable`, `create_alert`, `context_sources` conventions.
- `tests/synthetic/eks/run_suite.py` — scoring + trajectory pattern to mirror.

---

### Series cross-links

- [Part 1/5 — Provider/transport adapter failures (#1509)](https://github.com/Tracer-Cloud/opensre/issues/1509)
- [Part 2/5 — Long-running/agentic reliability (#1510)](https://github.com/Tracer-Cloud/opensre/issues/1510)
- [Part 3/5 — Multi-agent orchestration + memory (#1511)](https://github.com/Tracer-Cloud/opensre/issues/1511)
- [Part 4/5 — Determinism, approval, audit, RBAC, credential proxy (#1512)](https://github.com/Tracer-Cloud/opensre/issues/1512)
- [Part 5/5 — Surface-sprawl meta-test + harness wiring (#1513)](https://github.com/Tracer-Cloud/opensre/issues/1513)

### PR DESCRIPTION
**Pull request by muddlebee**
Fixes #1510

#### Describe the changes you have made in this PR -
This PR implements the Hermes RCA synthetic + long-running test track requested in issue #1510, aligned with the existing Hermes direction from merged sample PR #1860.

What’s included:
- Added a new parallel synthetic suite at `tests/synthetic/hermes_rca/` (keeps existing `tests/synthetic/hermes/` untouched).
- Added deterministic fixture backend `tests/synthetic/mock_hermes_backend/backend.py` used by both synthetic and e2e harnesses.
- Added Hermes investigation evidence tools in `app/tools/HermesSessionEvidenceTool/`:
  - `get_hermes_session_log`
  - `get_hermes_message_history`
  - `get_hermes_kv_cache_state`
  - `get_hermes_runtime_state`
  - `get_hermes_cron_state`
  - `get_hermes_session_topology`
- Added long-running e2e harness and tests under `tests/e2e/hermes/long_running/`.
- Added `pytest` marker wiring + Make targets:
  - `make test-hermes-synthetic`
  - `make test-hermes`
- Updated docs (`docs/hermes.mdx`) to reference the new parallel suite.

Follow-up fixes in this PR after first live run:
- Expanded root-cause taxonomy to include Hermes-specific categories used by the scenarios:
  - `agent_state_corruption`, `agent_hang`, `delivery_hang`, `ghost_session`, `performance_degradation`
- Prompt now uses taxonomy from `app/types/root_cause_categories.py` instead of hard-coded legacy category list.
- Structured diagnosis schema guidance updated to use taxonomy source-of-truth.
- Legacy category parser made slightly more robust for noisy formatting.
- Fixture backend now returns structured `available=False` envelopes for missing evidence instead of throwing, to reduce brittle LLM-path failures.

### Demo/Screenshot for feature changes and bug fixes -
Terminal evidence:

- Live e2e run started with:
  - `LLM_PROVIDER=codex CODEX_MODEL=gpt-5.4-mini OPENSRE_REASONING_EFFORT=low OPENSRE_RUN_HERMES_E2E=1 uv run pytest tests/e2e/hermes/long_running -m e2e -v`

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**
- Problem solved: provide a dedicated Hermes RCA benchmark track with realistic failure modes and enforce evidence-backed categorization behavior.
- Alternatives considered: (1) mutate existing `tests/synthetic/hermes` in place, (2) add a new parallel suite. Chosen (2) to avoid regressions and allow incremental rollout.
- Why this approach: shared fixture backend + new evidence tools keeps synthetic/e2e consistent while making scenarios explicit and maintainable.
- Key components:
  - `tests/synthetic/hermes_rca/*`: scenarios, alert fixtures, answer keys, suite runner.
  - `tests/synthetic/mock_hermes_backend/backend.py`: deterministic evidence transport.
  - `app/tools/HermesSessionEvidenceTool/*`: investigation-facing evidence collection.
  - `tests/e2e/hermes/orchestrator.py`: invokes OpenSRE investigation path with Hermes fixture context.
  - taxonomy/prompt updates: ensure emitted categories and parser expectations are consistent.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by muddlebee** [comment]
@greptile review\n\nPlease review this draft for Hermes RCA synthetic/e2e additions, evidence-tool contracts, and taxonomy/category handling.

**Comment by muddlebee** [comment]
@greptile review, pls check updated PR description and code

**Comment by muddlebee** [comment]
Addressed current review findings and pushed fixes in f9a7e0e9 (protocol ellipsis stubs replaced with explicit pass bodies in Hermes backend protocol).\n\n@greptile review

**Comment by muddlebee** [comment]
@greptile review

**Comment by muddlebee** [comment]
@greptile review

**Comment by muddlebee** [comment]
@greptile review

**Comment by muddlebee** [comment]
@greptile review

**Comment by muddlebee** [comment]
@greptile review

**Comment by Davidson3556** [comment]
some test is failing @muddlebee

**Comment by muddlebee** [comment]
> some test is failing @muddlebee

related to mintlify, can be ignored

File: app/agent/result.py [hunk 1]
```diff
@@ -8,15 +8,16 @@
...
 logger = logging.getLogger(__name__)
 
 
 class _DiagnosisSchema(BaseModel):
     root_cause: str = Field(description="Concise root cause statement (2-3 sentences max)")
     root_cause_category: str = Field(
         description=(
-            "One of: database, infrastructure, code_bug, configuration, "
-            "network, performance, healthy, unknown"
+            f"Use exactly one category from this taxonomy:\n{render_prompt_taxonomy().strip()}"
```
Comment (Devesh36) [inline app/agent/result.py]: render_prompt_taxonomy() renders all categories including Hermes ones into _DiagnosisSchema.description. This schema is used for every investigation — not just Hermes. So a Postgres alert's parsing step will see agent_hang as a valid category. Pass a filtered set to render_prompt_taxonomy()

**Comment by Devesh36** [comment]
rest all looks good

**Comment by cerencamkiran** [comment]
Huge amount of work here @muddlebee 

Only one very small thought I had while reading `tests/synthetic/hermes_rca/010-compression-invalid-tool-ordering/hermes_message_history.json`: the issue description mentions comparing both pre- and post-compression message ordering, but I think the current fixture may only represent the corrupted/final ordering state.

Maybe one possible follow-up could be including the valid pre-compression ordering too (or explicitly separating before/after snapshots), so the RCA can reason about the invariant break itself instead of only seeing the malformed final sequence.

But overall this looks good to me 🙌

**Comment by muddlebee** [comment]
thank you @Devesh36 @cerencamkiran reviews have been adressed here [7391694](https://github.com/Tracer-Cloud/opensre/pull/2075/commits/73916945171596fcf5d0b4ade3ed37425218c45d)
